### ISSUE TITLE
Absorb `when` package

### ISSUE THREAD

**Issue by seaneagan** [dart-lang/async#1]
Any interest in merging the [when](https://github.com/dart-lang/when) package into this one?

cc @pq

**Comment by pq** [comment]
Works for me.  Up to @lrhn to decide how aligned it is with `async`.

**Comment by lrhn** [comment]
I'm not sure I'd want to include it in package:async.
The problem is that it mixes sync and async behavior, returning a future in some cases and a value in other cases. That's something I generally try to avoid - a function is either async or sync, but not both, because that makes it clear what you can expect as a result, and how you need to treat it.
If you end up with an async result in a sync context, you are *****ed.

So, I'd make it always return a Future, but then it's not really necessary because you can just do:

     new Future.sync(callback).then(onSuccess, onError: onError).whenComplete(whenComplete)

(omittting whatever parameters you don't use).

**Comment by seaneagan** [comment]
The purpose of `when` is not to expose a mixed sync/async interface, but rather to implement two separate (though similar) sync and async interfaces to the same functionality without duplicating all the logic between both interfaces.

For example, it's [used to implement](https://github.com/dart-lang/which/blob/master/lib/src/which_impl.dart) `which` and `whichSync` in the `which` package:

It could potentially be built on top of an API in the `async.results` library which is the same, except takes a single `onResult` callback instead of separate `onSuccess`, `onError`, `onComplete` callbacks.

It is perfectly fine as a standalone package, but since it's pretty small, and async-related, I thought it might make sense to merge here.  Feel free to close if you disagree.

**Comment by lrhn** [comment]
I can see the lure of abstracting over combining operations without requiring them to be either sync or async - in such a way that you *can* get a synchronous result.
It's basically an abstraction over the operatons on futures, which is not a style I would generally recommend writing in. 

I'm worried that it requires you to be very careful no to introduce asynchrony accidentally, because there is no guard between sync and async operations. Sync operations embed easily into async, but the other way doesn't. If the onSucces function returns a Future, then your function is essentially asynchronous, even if your original computation isn't. 
Maybe you could have two functions: whenSync and whenAsync, and pass the correct version as an argument to your algorithm, that would seem safer. It's still a very specialized operation.

If abstrating over combining either sync or async operations was the goal, I think I would rather define a "synchronous future-like thingie", perhaps named Result, with similar then/catchError/whenComplete methods, and then build from that - possibly inside a helper class with a type parameter set to either Future or Result to ensure that I stay sync when I want that.

Then you could do the same logic:

class Helper<T> {
  T operation(T input) {
     var untypedInput = input;
     // Operations being combined should *not* be async themselves, which is still
     // pretty restrictive.
     return untypedInput.then((v) => v.map(something), onError: (e, s) => ...)
                                    .whenComplete(() { foo.close(); });
  }
}

and do either  
   new Helper<Future>().operation(asyncInput)
or
   new Helper<Result>().operation(syncInput).value;  // .value executes the built-up operation.



In short, there are a many possible helper functions that may be occasionally useful, but we can't include them all. What we include should preferably be generally useful, and generalize to problems we haven't had yet (which is obviously hard to predict) :). 
I think the when` function is too simple to generally useful - even if it exactly matches your use-case in `where`, which is cool. And then it breaks one of the design goals that I try to folllow: never have the same function return both sync and async results.
